File: app/src/main/java/de/karzek/diettracker/presentation/main/settings/dialog/manipulateMeal/ManipulateMealDialog.java
package de.karzek.diettracker.presentation.main.settings.dialog.manipulateMeal;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatDialogFragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Locale;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.karzek.diettracker.R;
import de.karzek.diettracker.presentation.util.Constants;
import de.karzek.diettracker.presentation.util.StringUtils;

/**
 * Created by MarjanaKarzek on 10.07.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 10.07.2018
 */
public class ManipulateMealDialog extends AppCompatDialogFragment {

    @BindView(R.id.meal_title)
    EditText mealTitle;

    @BindView(R.id.dialog_start_hour)
    EditText startHourField;
    @BindView(R.id.dialog_start_minute)
    EditText startMinuteField;
    @BindView(R.id.dialog_end_hour)
    EditText endHourField;
    @BindView(R.id.dialog_end_minute)
    EditText endMinuteField;

    @BindView(R.id.dialog_action_dismiss)
    Button dismiss;
    @BindView(R.id.dialog_action_save)
    Button save;

    private int id = -1;
    private String originalMealTitle;
    private String originalStartTime;
    private String originalEndTime;

    private Calendar startTimeCalendar = Calendar.getInstance();
    private Calendar endTimeCalendar = Calendar.getInstance();

    private SimpleDateFormat databaseTimeFormat = new SimpleDateFormat("HH:mm:ss", Locale.GERMANY);

    private AddMealFromDialogListener addListener;
    private SaveMealFromDialogListener saveListener;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.dialog_add_meal, container, false);

        ButterKnife.bind(this, view);

        Bundle bundle = this.getArguments();
        if (bundle != null) {
            id = bundle.getInt("mealId");
            originalMealTitle = bundle.getString("mealTitle");
            originalStartTime = bundle.getString("startTime");
            originalEndTime = bundle.getString("endTime");

            setupFieldsWithPreselection();
        } else {
            setupDefaultTime();
        }

        setupTimeChangeListener();

        dismiss.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                dismiss();
                addListener = null;
                saveListener = null;
            }
        });

        save.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(inputFieldsValid()) {
                    dismiss();
                    if(id == Constants.INVALID_ENTITY_ID)
                        addListener.addMealInDialogClicked(mealTitle.getText().toString(), databaseTimeFormat.format(startTimeCalendar.getTime()), databaseTimeFormat.format(endTimeCalendar.getTime()));
                    else
                        saveListener.saveMealInDialogClicked(id, mealTitle.getText().toString(), databaseTimeFormat.format(startTimeCalendar.getTime()), databaseTimeFormat.format(endTimeCalendar.getTime()));
                    addListener = null;
                    saveListener = null;
                } else {
                    showInvalidFieldsError();
                }
            }
        });

        return view;
    }

    private void setupFieldsWithPreselection() {
        mealTitle.setText(originalMealTitle);

        try {
            startTimeCalendar.setTime(databaseTimeFormat.parse(originalStartTime));
            endTimeCalendar.setTime(databaseTimeFormat.parse(originalEndTime));
        } catch (ParseException e) {
            e.printStackTrace();
            setupDefaultTime();
        }

        startHourField.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.HOUR_OF_DAY)));
        startMinuteField.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.MINUTE)));

        endHourField.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.HOUR_OF_DAY)));
        endMinuteField.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.MINUTE)));
    }

    private void setupTimeChangeListener() {
        startHourField.setOnEditorActionListener(new EditText.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    clearFocusOfView(startHourField);
                    startTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(startHourField.getText().toString()));
                }
                return true;
            }
        });
        startHourField.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                startTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(startHourField.getText().toString()));
            }
        });

        startMinuteField.setOnEditorActionListener(new EditText.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    clearFocusOfView(startMinuteField);
                    startTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(startMinuteField.getText().toString()));
                }
                return true;
            }
        });
        startMinuteField.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                startTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(startMinuteField.getText().toString()));
            }
        });

        endHourField.setOnEditorActionListener(new EditText.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    clearFocusOfView(endHourField);
                    endTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(endHourField.getText().toString()));
                }
                return true;
            }
        });
        endHourField.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                endTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(endHourField.getText().toString()));
            }
        });

        endMinuteField.setOnEditorActionListener(new EditText.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    clearFocusOfView(endMinuteField);
                    endTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(endMinuteField.getText().toString()));
                }
                return true;
            }
        });
        endMinuteField.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View view, boolean b) {
                endTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(endMinuteField.getText().toString()));
            }
        });
    }

    public void clearFocusOfView(EditText view) {
        InputMethodManager imm = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        view.clearFocus();
    }

    private void setupDefaultTime() {
        startTimeCalendar.set(Calendar.HOUR_OF_DAY, 8);
        startTimeCalendar.set(Calendar.MINUTE, 0);
        startTimeCalendar.set(Calendar.SECOND, 0);

        endTimeCalendar.set(Calendar.HOUR_OF_DAY, 10);
        endTimeCalendar.set(Calendar.MINUTE, 0);
        endTimeCalendar.set(Calendar.SECOND, 0);

        startHourField.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.HOUR_OF_DAY)));
        startMinuteField.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.MINUTE)));

        endHourField.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.HOUR_OF_DAY)));
        endMinuteField.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.MINUTE)));
    }

    private boolean inputFieldsValid() {
        return !mealTitle.getText().toString().equals("") && !startTimeCalendar.after(endTimeCalendar);
    }

    private void showInvalidFieldsError() {
        if (mealTitle.getText().toString().equals(""))
            mealTitle.setError(getString(R.string.error_message_missing_meal_title));
        if (startTimeCalendar.after(endTimeCalendar))
            Toast.makeText(getContext(), R.string.error_message_invalid_start_end_time, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            addListener = (AddMealFromDialogListener) getParentFragment();
            saveListener = (SaveMealFromDialogListener) getParentFragment();
        } catch (ClassCastException e) {
            throw new ClassCastException("fragment must implement AddMealFromDialogListener and SaveMealFromDialogListener");
        }
    }

    public interface AddMealFromDialogListener {
        void addMealInDialogClicked(String name, String startTime, String endTime);
    }

    public interface SaveMealFromDialogListener {
        void saveMealInDialogClicked(int id, String name, String startTime, String endTime);
    }
}

File: app/src/main/java/de/karzek/diettracker/presentation/main/settings/dialog/manipulateMeal/ManipulateMealDialog.kt
package de.karzek.diettracker.presentation.main.settings.dialog.manipulateMeal

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import android.widget.EditText
import android.widget.Toast
import androidx.appcompat.app.AppCompatDialogFragment
import de.karzek.diettracker.R
import de.karzek.diettracker.presentation.util.Constants
import de.karzek.diettracker.presentation.util.StringUtils
import kotlinx.android.synthetic.main.dialog_add_meal.*
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Created by MarjanaKarzek on 10.07.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 10.07.2018
 */
class ManipulateMealDialog : AppCompatDialogFragment() {

    private var currentId = -1
    private var originalMealTitle: String? = null
    private var originalStartTime: String? = null
    private var originalEndTime: String? = null

    private val startTimeCalendar = Calendar.getInstance()
    private val endTimeCalendar = Calendar.getInstance()

    private val databaseTimeFormat = SimpleDateFormat("HH:mm:ss", Locale.GERMANY)

    private var addListener: AddMealFromDialogListener? = null
    private var saveListener: SaveMealFromDialogListener? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.dialog_add_meal, container, false)

        val bundle = this.arguments
        if (bundle != null) {
            currentId = bundle.getInt("mealId")
            originalMealTitle = bundle.getString("mealTitle")
            originalStartTime = bundle.getString("startTime")
            originalEndTime = bundle.getString("endTime")

            setupFieldsWithPreselection()
        } else {
            setupDefaultTime()
        }

        setupTimeChangeListener()

        dialog_action_dismiss.setOnClickListener {
            dismiss()
            addListener = null
            saveListener = null
        }

        dialog_action_save.setOnClickListener {
            if (inputFieldsValid()) {
                dismiss()
                if (currentId == Constants.INVALID_ENTITY_ID)
                    addListener!!.addMealInDialogClicked(meal_title.text.toString(), databaseTimeFormat.format(startTimeCalendar.time), databaseTimeFormat.format(endTimeCalendar.time))
                else
                    saveListener!!.saveMealInDialogClicked(currentId, meal_title.text.toString(), databaseTimeFormat.format(startTimeCalendar.time), databaseTimeFormat.format(endTimeCalendar.time))
                addListener = null
                saveListener = null
            } else {
                showInvalidFieldsError()
            }
        }

        return view
    }

    private fun setupFieldsWithPreselection() {
        meal_title.setText(originalMealTitle)

        try {
            startTimeCalendar.time = databaseTimeFormat.parse(originalStartTime!!)!!
            endTimeCalendar.time = databaseTimeFormat.parse(originalEndTime!!)!!
        } catch (e: ParseException) {
            e.printStackTrace()
            setupDefaultTime()
        }

        dialog_start_hour.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.HOUR_OF_DAY)))
        dialog_start_minute.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.MINUTE)))

        dialog_end_hour.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.HOUR_OF_DAY)))
        dialog_end_minute.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.MINUTE)))
    }

    private fun setupTimeChangeListener() {
        dialog_start_hour.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                clearFocusOfView(dialog_start_hour)
                startTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(dialog_start_hour.text.toString()))
            }
            true
        }
        dialog_start_hour.onFocusChangeListener = View.OnFocusChangeListener { _, _ -> startTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(dialog_start_hour.text.toString())) }

        dialog_start_minute.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                clearFocusOfView(dialog_start_minute)
                startTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(dialog_start_minute.text.toString()))
            }
            true
        }

        dialog_start_minute.onFocusChangeListener = View.OnFocusChangeListener { _, _ -> startTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(dialog_start_minute.text.toString())) }

        dialog_end_hour.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                clearFocusOfView(dialog_end_hour)
                endTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(dialog_end_hour.text.toString()))
            }
            true
        }
        dialog_end_hour.onFocusChangeListener = View.OnFocusChangeListener { _, _ -> endTimeCalendar.set(Calendar.HOUR_OF_DAY, Integer.valueOf(dialog_end_hour.text.toString())) }

        dialog_end_minute.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                clearFocusOfView(dialog_end_minute)
                endTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(dialog_end_minute.text.toString()))
            }
            true
        }
        dialog_end_minute.onFocusChangeListener = View.OnFocusChangeListener { _, _ -> endTimeCalendar.set(Calendar.MINUTE, Integer.valueOf(dialog_end_minute.text.toString())) }
    }

    private fun clearFocusOfView(view: EditText) {
        val imm = context!!.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(view.windowToken, 0)
        view.clearFocus()
    }

    private fun setupDefaultTime() {
        startTimeCalendar.set(Calendar.HOUR_OF_DAY, 8)
        startTimeCalendar.set(Calendar.MINUTE, 0)
        startTimeCalendar.set(Calendar.SECOND, 0)

        endTimeCalendar.set(Calendar.HOUR_OF_DAY, 10)
        endTimeCalendar.set(Calendar.MINUTE, 0)
        endTimeCalendar.set(Calendar.SECOND, 0)

        dialog_start_hour.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.HOUR_OF_DAY)))
        dialog_start_minute.setText(StringUtils.formatInt(startTimeCalendar.get(Calendar.MINUTE)))

        dialog_end_hour.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.HOUR_OF_DAY)))
        dialog_end_minute.setText(StringUtils.formatInt(endTimeCalendar.get(Calendar.MINUTE)))
    }

    private fun inputFieldsValid(): Boolean {
        return meal_title.text.toString() != "" && !startTimeCalendar.after(endTimeCalendar)
    }

    private fun showInvalidFieldsError() {
        if (meal_title.text.toString() == "")
            meal_title.error = getString(R.string.error_message_missing_meal_title)
        if (startTimeCalendar.after(endTimeCalendar))
            Toast.makeText(context, R.string.error_message_invalid_start_end_time, Toast.LENGTH_SHORT).show()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        try {
            addListener = parentFragment as AddMealFromDialogListener?
            saveListener = parentFragment as SaveMealFromDialogListener?
        } catch (e: ClassCastException) {
            throw ClassCastException("fragment must implement AddMealFromDialogListener and SaveMealFromDialogListener")
        }

    }

    interface AddMealFromDialogListener {
        fun addMealInDialogClicked(name: String, startTime: String, endTime: String)
    }

    interface SaveMealFromDialogListener {
        fun saveMealInDialogClicked(id: Int, name: String, startTime: String, endTime: String)
    }
}
